source/common/http/http2/codec_impl.cc

```diff
@@ -1136,9 +1153,12 @@ ClientConnectionImpl::ClientConnectionImpl(
                      max_response_headers_count),
       callbacks_(callbacks) {
   ClientHttp2Options client_http2_options(http2_options);
-  nghttp2_session_client_new2(&session_, http2_callbacks_.callbacks(), base(),
-                              client_http2_options.options());
-  sendSettings(http2_options, true);
+  Nghttp2SessionFactory* session_factory =
+      Registry::FactoryRegistry<Nghttp2SessionFactory>::getFactory("envoy.http2.session");
```

Comment:
Generally we only use this type of factory for user configuration, and configuration that is not used in hi perf paths. Would it be possible to not use the static factory system for this? The simplest way of dealing with this (which we do in a lot of other places) is to just have a factory interface and then have a prod/default impl which is used at prod construction points and that can be overridden in tests?

Reply:
Right, I was concerned about the potential perf hit of reusing this, but it's currently using an O(1) lookup to find the registered factory. With that said, I understand your concern that this could change in the future and this performance sensitive path could be overlooked.

I looked for other abstractions for static factories and couldn't find any (I did find quite a few non static use cases); I'll create a (simple) new one for this, or if I've missed one, please let me know.

Or are you suggesting extending the interface of the `ConnectionImpl` to take an nghttp2_session factory as a constructor arg?

Reply:
> Or are you suggesting extending the interface of the ConnectionImpl to take an nghttp2_session factory as a constructor arg?

No I was just suggesting a basic factory pattern. I don't think we have anything boilerplate but we do this in a bunch of places beyond the static factory system (LMK if you want a more direct reference to something similar and I can dig one up.)

Reply:
Ok. I'm working on this; if you can dig up one of the existing examples I would certainly appreciate that as well to validate my approach. Thanks!

Reply:
One pattern is to do the base class and then have a "prod" override. So something like `CodecClientProd` or `ProdHttpHealthCheckerImpl`. This might not work due to the virtual constructor problem. In the other case I would make a factory and use it where an object needs to be created, so something more like `TapConfigFactory`.

Reply:
> In the other case I would make a factory and use it where an object needs to be created, so something more like `TapConfigFactory`.

Right. The open question right now for me is how to inject the factory from the test. So far two options have been ruled out: 1) use `FactoryRegistry`, and 2) add a constructor argument.

That seems to leave two other options: 3) create a simpler static registry akin to `Registry::FactoryRegistry`, or 4) use a singleton (e.g., `InjectableSingleton`) for the factory, with the downside that initialization of the singleton needs to happen outside the `ConnectionImpl`s translation unit.

Reply:
If constructor arg is the easiest that is fine also. Whatever you think is best.

Reply:
I refactored to simply pass a factory as a constructor arg. This is the simplest approach and matches the style used elsewhere.